core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -681,4 +642,120 @@ private void doRelease() {
             }
         }
     }
+
+    /**
+     * A variant of {@link CompletableFuture} that keeps its completion handlers into a separate list.
+     * This yields better performance than {@link CompletableFuture#handle(BiFunction)} as the number of
+     * added handlers increases because it does not create a long linked list with extra wrappers, which is
+     * especially beneficial for Java 8 which suffers a huge performance hit when complementing a future with
+     * a deep stack. See {@code cleanStack()} in Java 8 {@link CompletableFuture} for more information.
+     */
+    private final class ChannelAcquisitionFuture extends CompletableFuture<PooledChannel> {
+
+        @Nullable
+        private List<Consumer<PooledChannel>> notifications;
```

Comment:
We could consider inlining a single notification since it seems like there will be two common cases, only one notification, or many notifications. But maybe premature

Reply:
Done. PTAL again :smile: